### PR TITLE
Joint-Wasserstein-GP Implementation

### PR DESCRIPTION
**Pull request by rl-max**
**Feature added:**

1. JointGAN version of Wasserstein Loss (use Gradient Penalty) 
2. LSGAN new coding scheme (-1, 0, 1) [**Deprecated**]

Currently just open for Review

### CODE REVIEW

File: src/utils/losses.py [hunk 1]
```diff
@@ -222,16 +222,21 @@ def g_ls_relative(d_logit_fake, DDP, d_logit_real=None):
...
+def g_wasserstein_relative(d_logit_fake, DDP, d_logit_real=None):
+    if d_logit_real is None:
+        return g_wasserstein(d_logit_fake, DDP)
+    else:
+        return torch.mean(d_logit_real - d_logit_fake)
+
+
+def d_ls_relative(d_logit_real, d_logit_fake,  DDP):
+    d_loss = 0.5 * (d_logit_real - 1.0)**2 + \
+             0.5 * (d_logit_fake + 1.0)**2
```
Comment (krenerd) [inline src/utils/losses.py:238]:
```suggestion
    d_loss = 0.5 * (d_logit_real - 1.0) ** 2 + \
                 0.5 * (d_logit_fake + 1.0) ** 2
```

File: src/utils/losses.py [hunk 2]
```diff
@@ -222,16 +222,21 @@ def g_ls_relative(d_logit_fake, DDP, d_logit_real=None):
     if d_logit_real is None:
         return g_ls(d_logit_fake, DDP)
     else:
-        gen_loss = 0.5 * (d_logit_fake - torch.ones_like(d_logit_fake))**2 + 0.5 * (d_logit_real)**2
```
Comment (sieu-n) [inline src/utils/losses.py]:
```suggestion
        gen_loss = 0.5 * (d_logit_fake - torch.ones_like(d_logit_fake))**2 + 0.5 * (d_logit_real)**2
```

File: src/utils/losses.py [hunk 3]
```diff
@@ -222,16 +222,21 @@ def g_ls_relative(d_logit_fake, DDP, d_logit_real=None):
...
+    if d_logit_real is None:
+        return g_wasserstein(d_logit_fake, DDP)
+    else:
+        return torch.mean(d_logit_real - d_logit_fake)
+
+
+def d_ls_relative(d_logit_real, d_logit_fake,  DDP):
+    d_loss = 0.5 * (d_logit_real - 1.0)**2 + \
+             0.5 * (d_logit_fake + 1.0)**2
+    return d_loss.mean()
```
Comment (sieu-n) [inline src/utils/losses.py]:
```suggestion
```

**Review by sieu-n — CHANGES_REQUESTED** [review]
LGTM, though remove lsgan related stuff!

File: src/config.py [hunk 4]
```diff
@@ -438,7 +439,8 @@ def define_losses(self):
                 "rgan": losses.d_rgan,
                 "ragan": losses.d_ragan,
                 "vanilla_relativistic": losses.d_vanilla,
-                "least_square_relativistic": losses.d_ls,
+                "least_square_relativistic": losses.d_ls_relative,
```
Comment (sieu-n) [inline src/config.py]:
```suggestion
                "least_square_relativistic": losses.d_ls,
```

**Review by sieu-n — CHANGES_REQUESTED** [review]
review 2 forgot this

File: src/utils/losses.py [hunk 5]
```diff
@@ -222,16 +222,17 @@ def g_ls_relative(d_logit_fake, DDP, d_logit_real=None):
     if d_logit_real is None:
         return g_ls(d_logit_fake, DDP)
     else:
-        gen_loss = 0.5 * (d_logit_fake - torch.ones_like(d_logit_fake))**2 + 0.5 * (d_logit_real)**2
+        gen_loss = 0.5 * (d_logit_fake)**2 + 0.5 * (d_logit_real)**2
```
Comment (sieu-n) [inline src/utils/losses.py]:
```suggestion
        gen_loss = 0.5 * (d_logit_fake - torch.ones_like(d_logit_fake))**2 + 0.5 * (d_logit_real)**2
```

**Review by sieu-n — CHANGES_REQUESTED** [review]
.

**Review by sieu-n — APPROVED** [review]
LGTM
